File: .ipynb_checkpoints/functionGbound-checkpoint.jl
function gx_bound(c, c_g, edge)
    
    #println(f,"Current CELL's LB = ", c_L)
    #println(f,"Current CELL's UB = ", c_U)
    #println(f, "Current interdiction x = ", x_now)
    start_node = edge[:,1]
    end_node = edge[:,2]

    no_node = max(maximum(start_node), maximum(end_node) )
    no_link = length(start_node)


    function getShortestX(state, start_node, end_node, origin, destination)
        _x = zeros(Int, length(start_node))
        _path = enumerate_paths(state, destination)

        for i=1:length(_path)-1
            _start = _path[i]
            _end = _path[i+1]

            for j=1:length(start_node)
                if start_node[j]==_start && end_node[j]==_end
                _x[j] = 1
                break
                end
            end

        end
        _x
    end


    graph = SimpleDiGraph(no_node)
    distmx = Inf*ones(no_node, no_node)

    # Adding links to the graph
    for i=1:no_link
        add_edge!(graph, start_node[i], end_node[i])
        distmx[start_node[i], end_node[i]] = c_g[i]
    end

    # Run Dijkstra's Algorithm from the origin node to all nodes
    state = dijkstra_shortest_paths(graph, origin, distmx)
    label = state.dists
    pred = state.parents
    b_arc = ""
    
    for i = 1: length(state.parents)
        if state.parents[i] != 0 
            b_arc = string(b_arc, "(", state.parents[i], ",", i, ")")
        end
    end
    
    # Retrieving the shortest path
    path = enumerate_paths(state, destination)
    
    #parents = LightGraphs.DijkstraState(state, destination)

    # Retrieving the 'x' variable in a 0-1 vector
    y = getShortestX(state, start_node, end_node, origin, destination)
    #println(f,"y vector:", y)
    
    gx = sum(c_g[i]*y[i] for i = 1:no_link)    
    SP = sum(c[i]*y[i] for i = 1:length(c))
    T = Int64[]
    for i = 1:Len
        if pred[edge[i,2]] == edge[i,1]
            push!(T, 1)
        else
            push!(T, 0)
        end
    end

    #println("Minimum Spanning Tree: ", T)
    #println("Shortest path y = ", y)
    #println("Indices of shortest path (edge) = ", y_index)
    #println("Nodes visited =", path)
    #println("Minimum Spanning Tree =", pred)
    #println("Node label =" ,label)

    return y, gx, SP, T, pred, label, path
end

File: PythonConversion/.ipynb_checkpoints/functionGbound-checkpoint.py
import networkx as nx
import numpy as np

def gx_bound(c, c_g, edge,origin,destination):
    G = nx.DiGraph()
    for _e in range(len(c)):
        # print("\t", edge[_e])
        G.add_edge(edge[_e,0],edge[_e,1], weight=c_g[_e])
    # distmx = np.full((no_node, no_node), np.inf)
    # Find the shortest paths and lengths from the source node to all other nodes
    shortest_paths_lengths = nx.single_source_dijkstra_path_length(G, source=origin, weight='weight')
    # print(G)
#     # Get the list of predecessors for each node
#     predecessors = nx.predecessor(G, source=origin)

#     # Print the results
#     for node, predecessor_list in predecessors.items():
#         print(f"Node {node}: Predecessors {predecessor_list}")
    # print(start_node)
    # print(end_node)
    # print(distmx)
    # print(c_g)
    # print(no_link)
    # Adding links to the graph
    # Print the results
    # print(shortest_paths_lengths[1])
    label = list(shortest_paths_lengths.values())
    # print("label ", label)
    # print(f"Shortest paths from node {origin} to all other nodes:")
    # for node, path_length in shortest_paths_lengths.items():
    #     print(f"To node {node}: Length = {path_length}, Path = {nx.shortest_path(G, source=origin, target=node, weight='weight')}")
    
    
    y = np.zeros(len(G.edges))
    shortest_path=nx.shortest_path(G, source=origin, target=destination, weight='weight')
    # print("shortest_path ", shortest_path)
    for i in range(len(shortest_path) - 1):
        start_node = shortest_path[i]
        end_node = shortest_path[i + 1]
        # print("\t", start_node, "\t", end_node)
        # print(list(G.edges))
        # edge_index = list(G.edges).index((start_node, end_node)) #this rearranges the elements - don't do this
        a = np.array([start_node, end_node])
        edge_index = np.where((edge==a).all(1))[0]
        # print("\t edge_index ", edge_index)
        y[edge_index] = 1
    # print("y = ", y)
    gx = np.sum(np.array(c_g) * y)
    SP = np.sum(np.array(c) * y)
    
    # T = [1 if pred[edge[i, 1]] == edge[i, 0] else 0 for i in range(no_link)]

    return y, gx, SP, label, shortest_path
